### ISSUE TITLE
Import data from other platforms (Steam/GOG/etc)

### ISSUE THREAD

**Issue by mattman107** [h-quer/Slaanesh#47]
I'm not 100% sure if this falls into the Out of Scope section "Social media features", but I figured I would bring it up. 

As someone who a big amount of data across many different platforms and would like to use a tool like this to track it I find it difficult to start using a tool like this without something to help get me in the door. If there was a built in way for Slaanesh to search a steam account (or others) to scrape games/playtime that would be huge.

**Comment by h-quer** [comment]
Yes, that's definitely something I'd like to include eventually. There is a CSV import right now, so if you can get your list of Steam games as CSV (e.g. by exporting it from tools like Playnite) it's reasonably easy to get startet. Eventually, I'd like to be able to directly import from Steam. I've started looking into the Steam API for that already, but haven't made much progress so far.

Ideally, I'd like to extract the game name, game Steam ID (which I can then use to lookup the corresponding IGDB ID to add to the database), time last played, and also any Steam library collections / categories the game is in. If anyone happens to know the Steam API functionality to extract this information, I'd be very happy for a few pointers.

**Comment by mattman107** [comment]
I can't say that I know how to use the Steam api at all, but I am working on some other changes to Slaanesh (look forward to some stuff coming soon!). If I find some time I will try my hand at it.

**Comment by h-quer** [comment]
That does sound interesting, looking forward to it! Do you have a pointer as to the topic of the changes?

Life has been busy so my progress is rather slow, but I am currently working on the "database" update, enabling more and flexible columns for games and playthroughs.

**Comment by mattman107** [comment]
> Life has been busy so my progress is rather slow

I feel this. So true.

> Do you have a pointer as to the topic of the changes?

I am currently working on extras to backing up the database files.
1. Backup and download from the browser.
2. Automated backups.

>I am currently working on the "database" update, enabling more and flexible columns for games and playthroughs.

I was planning on doing something involving the databases but I haven't worked on it yet. Here was my idea:

So currently when you have multiple playthroughs for the same game it looks like this:
![image](https://github.com/user-attachments/assets/8ed6df7b-cbba-41ae-8629-19d5d636fda3)

They always display the same overall game status. In my case Celeste displays 'Complete'. The only difference is the date and any comments I could add for them. I was thinking about having a status on a per playthrough basis and also a playthrough type. The playthrough type would be a category for distinguishing between games you playedthrough just the main game, or the dlc, or played the game modded.

I thought that could be a cool statistic and way to sort playthroughs.

On a separate note, I have noticed that if you have too many platforms, the x-axis on this graph doesn't display correctly. Maybe it would work out better to have a separate page for each graph or something so these graphs fit nicely. Or maybe  allow the user to dynamically change graphs to display certain data points. I'm still letting this one cook.

On an even more separate note, it would be cool to export specific data points these graphs contain. So for example on the 'Platform stats' graph, if you could export all of the 'PC' games that you have or on the 'Yearly stats' graph if you could export all the games that you 'completed' in the year 2023.

**Comment by h-quer** [comment]
Thanks, I'll look into the axis of that graph!

Regarding database backups: They are just files in whatever directly gets mounted to the docker container. I've always considered backup functionality inside Slaanesh to be of little use, since these files will be gathered automatically by any backup solution digesting the mounted folder. But then again, offering the option doesn't do any harm, either. And a way to download the database (and the CSV export files) directly from within the UI would indeed be very useful, yes.

Regarding the game and playthrough status: I plan on completely re-working the way the "database" works right now, which would also enable the functionality you describe. The goal is to ultimately have both status and platform data in the playthrough table.

Platforms in the game table could then be a list of available platforms, from which the playthrough selects one on which the game was played.

Each playthrough can have its own status, e.g. first discarded, then completed, then at a later time mastered. I'm not sure how to best handle the game status, though. One option is to simply set it to "played" whenever a playthrough it recorded and then handling all the status stuff only with the pt data. Another option is to keep the game status as is, or even automatically set it to the "best" of all recorded pt status (e.g. if there is a discarded and a completed pt, the game overall counts as completed), and the pt status is an additional, optional information. In this case, the pt status could simply be displayed below the game status in the played table, if they differ.

Additionally, I want to add a few optional columns (incl. UI options to enable/disable them), e.g. version, playtime for playthoughs, rating for games).

I'm also thinking about using SQLite instead of simply dumping the Pandas dataframe into a feather file.

**Comment by mattman107** [comment]
Wow that sounds like a big re-write. When you meant re-write I assumed it would still be using the same Pandas dataframes. But this is essentially scrapping the old and re-creating/upgrading it into sqlite.

The reason I think it is cool to have more ways to backup/export the data files is because this is something I run in a docker container on a VM that I won't remote into often. So setting up a way to automatically create the .csv backups and using something else (like syncthing) to move the files to a remote location for external storage. But I also understand that some people might not want to go that crazy. That was why I thought allowing users to directly download from the GUI would be helpful.

**Comment by h-quer** [comment]
The new version now can import from GOG (which in turn can import from all major stores) and also pull additional data from Steam.
